### ISSUE TITLE
Refactor mutt_parse_hook()

### ISSUE THREAD

**Issue by flatcap** [neomutt/neomutt#2646]
[`mutt_parse_hook()`](https://github.com/neomutt/neomutt/blob/master/hook.c#L85) parses [19 different hook commands]():
- `account-hook`, `append-hook`, `charset-hook`, `close-hook`, `crypt-hook`,
  `fcc-hook`, `fcc-save-hook`, `folder-hook`, `iconv-hook`, `mbox-hook`,
  `message-hook`, `open-hook`, `reply-hook`, `save-hook`, `send-hook`,
  `send2-hook`, `shutdown-hook`, `startup-hook`, `timeout-hook`

This makes it quite difficult to read.

The hooks can be split into nine sets, based on the parameters that they take:

| [Hook](https://github.com/neomutt/neomutt/blob/master/mutt_commands.c#L51)            | Parameter | Parameter     | [Flags](https://github.com/neomutt/neomutt/blob/master/hook.h#L43)                                    |
| :-------------- | :-------- | :------------ | :--------------------------------------- |
| `charset-hook`  | alias     | charset       | `MUTT_CHARSET_HOOK`                      |
| `iconv-hook`    | charset   | local-charset | `MUTT_ICONV_HOOK`                        |
| `shutdown-hook` | command   |               | `MUTT_SHUTDOWN_HOOK \| MUTT_GLOBAL_HOOK` |
| `startup-hook`  | command   |               | `MUTT_STARTUP_HOOK \| MUTT_GLOBAL_HOOK`  |
| `timeout-hook`  | command   |               | `MUTT_TIMEOUT_HOOK \| MUTT_GLOBAL_HOOK`  |
| `message-hook`  | pattern   | command       | `MUTT_MESSAGE_HOOK`                      |
| `reply-hook`    | pattern   | command       | `MUTT_REPLY_HOOK`                        |
| `send-hook`     | pattern   | command       | `MUTT_SEND_HOOK`                         |
| `send2-hook`    | pattern   | command       | `MUTT_SEND2_HOOK`                        |
| `fcc-hook`      | pattern   | mailbox       | `MUTT_FCC_HOOK`                          |
| `fcc-save-hook` | pattern   | mailbox       | `MUTT_FCC_HOOK \| MUTT_SAVE_HOOK`        |
| `save-hook`     | pattern   | mailbox       | `MUTT_SAVE_HOOK`                         |
| `account-hook`  | regex     | command       | `MUTT_ACCOUNT_HOOK`                      |
| `folder-hook`   | regex     | command       | `MUTT_FOLDER_HOOK`                       |
| `crypt-hook`    | regex     | keyid         | `MUTT_CRYPT_HOOK`                        |
| `mbox-hook`     | regex     | mailbox       | `MUTT_MBOX_HOOK`                         |
| `append-hook`   | regex     | shell-command | `MUTT_APPEND_HOOK`                       |
| `close-hook`    | regex     | shell-command | `MUTT_CLOSE_HOOK`                        |
| `open-hook`     | regex     | shell-command | `MUTT_OPEN_HOOK`                         |

## Task 1

Split up `mutt_parse_hook()` into separate functions for each group of hooks.
This will lead to a small amount of duplicated code, but each function will be much easier to understand.

## Task 2

Examine the global variables that each function uses.
Where possible, move the new functions into libraries to reduce dependencies.
The functions should be moved into `{library}/commands.c`

## Task 3

Flag three old commands.

This is slighly unrelated, but create a new flag `MUTT_HOOK_DEPRECATED`
and add it to these commands:

- `pgp-hook`, `unvirtual-mailboxes`, `virtual-mailboxes`

This flag will be used in the future to deprecated old commands.

**Comment by jindraj** [comment]
There's one more occurrence of `mutt_parse_hook` in the code in init.c inside `mutt_get_hook_type()`. The function is called inside `mutt_parse_unhook()`. I don't fully understand what I'm supposed to do here.
https://github.com/neomutt/neomutt/blob/f84c4c4e39dc16248037a177265d2a2725ad3585/init.c#L695-L707

With my current code changes it still works as expected, because function `mutt_parse_hook()` still exists as it used to so it's registered in mutt_commands_array. But this part is crucial to get hook flags when unhook is called.

**Comment by flatcap** [comment]
This function is doing a lookup of a name, say `send-hook`, in the array of Commmands.
It's checking the `parse` member to avoid doing up to 78 string comparisons.

The simplest solution is to drop this function and use `mutt_command_get()` instead.

**Comment by klartext** [comment]
There seem to be at least some of the broken-down functions:
mutt_message_hook(), mutt_startup_shutdown_hook()?!
But they will not be called.
And it looks like this has been work in progress and then  stopped at some point and is unfinished now.

What is planned for this issue?
Can the two separate functions  be considered to be finished work (just not used so far)?

**Comment by flatcap** [comment]
> What is planned for this issue?

Needs doing.  Volunteer needed :heart:

**Comment by klartext** [comment]
The function is long... I need a larger screen for it... but I doubt to find one, that is large enough ;-)
Printed with 6-point font, the file would need 11 pages...

**Comment by flatcap** [comment]
> The function is long

Nah!  It's only 223 lines long -- it doesn't even make the top ten!

- 1087 `mutt_buffer_select_file()`
- 1013 `index_format_str()`
- 950 `main()`
- 850 `mutt_send_message()`
- 678 `mutt_enter_string_full()`
- 535 `mutt_expando_format()`
- 443 `mutt_rfc822_parse_line()`
- 440 `mutt_pager()`
- 378 `parse_set()`
- 377 `mbox_mbox_sync()`

And before I refactored the index and pager, there were worse!

- 2919 `mutt_index_menu()`
- 1849 `mutt_pager()`

**Comment by klartext** [comment]
WWW TTT FFF ;-)

**Comment by klartext** [comment]
The variable names are not realy telling: 
```mutt_parse_hook(struct Buffer *buf, struct Buffer *s, intptr_t data, struct Buffer *err)```

I think ```err``` and ```data``` are ok (even though a comment on ```data``` would be helpful, or maybe pick a better name).

But a variable of type ```Buffer``` with the name ```buf```, which stands for buffer, is not saying much.
And Buffer ```s``` also does not say much.

What is the meaning of these parameters?

**Comment by flatcap** [comment]
> The variable names are not realy telling:
> `mutt_parse_hook(struct Buffer *buf, struct Buffer *s, intptr_t data, struct Buffer *err)`

Yeah, I know.  They're historic names that I haven't quite got around to improving -- suggestions welcome.
However, they are consistent and they're documented.

In NeoMutt there are nearly 40 APIs - groups of functions with the same prototype.
In each case, we've created a typedef for the prototype and documented it once.
For each function, you'll see two keywords: **Implements** and **@ingroup**.

```c
/**
 * mutt_parse_hook - Parse the 'hook' family of commands - Implements Command::parse() - @ingroup command_parse
 *
 * This is used by 'account-hook', 'append-hook' and many more.
 */
enum CommandResult mutt_parse_hook(struct Buffer *buf, struct Buffer *s, intptr_t data, struct Buffer *err)
```

The Implements is the name of the API.
If you select `Command` and hit <kbd>Ctrl-]</kbd> (in vim), it'll take you to the definition.
The definition includes the `parse()` function, which is fully documented.

The @ingroup creates a name that Doxygen can index.
There's an [API tab](https://neomutt.org/code/modules.html) in the web Code Docs.

### History

Early on, I noticed that 17 functions implemented string formatting code.
Each was called from the same place, but they all had different parameter names and were used differently.
I unified all the functions, and created one clearly documented prototype.

https://github.com/neomutt/neomutt/blob/23b03a69f7c4131cf603af105a389a9173b5272f/hdrline.c#L441-L444

https://github.com/neomutt/neomutt/blob/23b03a69f7c4131cf603af105a389a9173b5272f/format_flags.h#L40-L67

**Comment by klartext** [comment]
Thanks, this is very helpful.
Especially the link to the API docs opens a hole new world ;-)

**Comment by klartext** [comment]
In
https://github.com/neomutt/neomutt/pull/2686
I found a 'hooks compilation'.
I can use that as testcase for parsable hooks.
The error buffer does not report errors with these hooks.
So that's nice to have these hooks for testing, when refactoring the code.
Possibly it makes sense to put that file somewhere into the source tree - as testing ressource.

Are there also hook-examples which will be rejected by the parser - with well known behaviour?
(I could try to mess the settings up randomly; but maybe there is a more systematic way for this.)

And are there also hook-definitions that *should* work, but the hook-parser parses them wrong or even crashes with them?

Even though at the moment I only want to do the refactoring, not repairing stuff, the good and the bad input is helpful to get all if-statements triggered. So if there might somewhere be a collection of parser-input files, please let me know.

**Comment by klartext** [comment]
...maybe adding tests to the test directory would be a way to go here...

**Comment by klartext** [comment]
How was that nice graph created?

**Comment by flatcap** [comment]
egypt, vim-fu, graphviz

First, you need to build NeoMutt with some extra flags
```sh
export EXTRA_CFLAGS="-fdump-rtl-expand"
./configure ...
make
```

That'll generate a lot of `*.expand` files.

Next, use the `egypt` tool.  We have a [slightly tweaked mirror](https://github.com/neomutt/egypt).
That turns the data in the `.expand` files into `.gv` (GraphViz 'dot' format).

```sh
egypt hook.c.*.expand > hook.gv
```

You can include lots of files on the command line.
I tend to include too many, then use vim-fu to filter out the junk.

Next, [GraphViz](https://graphviz.org/) has a couple of tools:
- `display hook.gv` (a viewer for `.gv` files)
- `dot -Tsvg hook.gv > hook.svg` (convert it to SVG)

**Comment by klartext** [comment]
Thanks for the detailed description of the documentation graphics. :-)

It looks like in the whole hook-parsing thing a lot of global values - indirectly used by the called functions - are involved.
This makes things harder to change. But I could do some refactoring so far.

One thing I'm just curious about is this condition to set ```use_regex``` to ```false``` here:
https://github.com/neomutt/neomutt/blob/master/hook.c#L106
Why is there also the condition ```folder_or_mbox``` included?

Is this intended? Does the ```-noregex``` only have to be in effect for folder and mbox hooks? (If so: why?)

BTW: looking for 'noregex'  here: https://neomutt.org/
did not give anything as result.

**Comment by klartext** [comment]
Reordering the hooks/parameters/flags table I would come with this suggestion:

```
Hook           Parameter      Parameter      Flags                  
-------------  -------------  -------------  -----------------
startup-hook   ./.            command        MUTT_STARTUP_HOOK   |  MUTT_GLOBAL_HOOK
shutdown-hook  ./.            command        MUTT_SHUTDOWN_HOOK  |  MUTT_GLOBAL_HOOK
timeout-hook   ./.            command        MUTT_TIMEOUT_HOOK   |  MUTT_GLOBAL_HOOK
message-hook   pattern        command        MUTT_MESSAGE_HOOK      
send-hook      pattern        command        MUTT_SEND_HOOK         
send2-hook     pattern        command        MUTT_SEND2_HOOK        
reply-hook     pattern        command        MUTT_REPLY_HOOK        
account-hook   regex          command        MUTT_ACCOUNT_HOOK      
folder-hook    regex          command        MUTT_FOLDER_HOOK       
fcc-hook       pattern        mailbox        MUTT_FCC_HOOK          
fcc-save-hook  pattern        mailbox        MUTT_FCC_HOOK       |  MUTT_SAVE_HOOK
save-hook      pattern        mailbox        MUTT_SAVE_HOOK         
mbox-hook      regex          mailbox        MUTT_MBOX_HOOK         
append-hook    regex          shell-command  MUTT_APPEND_HOOK       
close-hook     regex          shell-command  MUTT_CLOSE_HOOK        
open-hook      regex          shell-command  MUTT_OPEN_HOOK         
charset-hook   alias          charset        MUTT_CHARSET_HOOK      
iconv-hook     local-charset  charset        MUTT_ICONV_HOOK        
crypt-hook     regex          keyid          MUTT_CRYPT_HOOK        
```

Or with switched rows of parameters:

```Hook            Parameter      Parameter     Flags                  
-------------   -------------  ------------- -----------------
startup-hook    command        ./.           MUTT_STARTUP_HOOK   |  MUTT_GLOBAL_HOOK
shutdown-hook   command        ./.           MUTT_SHUTDOWN_HOOK  |  MUTT_GLOBAL_HOOK
timeout-hook    command        ./.           MUTT_TIMEOUT_HOOK   |  MUTT_GLOBAL_HOOK
message-hook    command        pattern       MUTT_MESSAGE_HOOK
send-hook       command        pattern       MUTT_SEND_HOOK
send2-hook      command        pattern       MUTT_SEND2_HOOK
reply-hook      command        pattern       MUTT_REPLY_HOOK
account-hook    command        regex         MUTT_ACCOUNT_HOOK
folder-hook     command        regex         MUTT_FOLDER_HOOK
fcc-hook        mailbox        pattern       MUTT_FCC_HOOK
fcc-save-hook   mailbox        pattern       MUTT_FCC_HOOK       |  MUTT_SAVE_HOOK
save-hook       mailbox        pattern       MUTT_SAVE_HOOK
mbox-hook       mailbox        regex         MUTT_MBOX_HOOK
append-hook     shell-command  regex         MUTT_APPEND_HOOK
close-hook      shell-command  regex         MUTT_CLOSE_HOOK
open-hook       shell-command  regex         MUTT_OPEN_HOOK
charset-hook    charset        alias         MUTT_CHARSET_HOOK
iconv-hook      charset        local-charset MUTT_ICONV_HOOK
crypt-hook      keyid          regex         MUTT_CRYPT_HOOK
```
(At the moment I prefer the first table.)

**Comment by flatcap** [comment]
> One thing I'm just curious about is this condition to set `use_regex` to `false`

Yes, `-noregex` is a new-ish feature.

By default `folder-hook` and `mbox-hook` take a regex, e.g. `.*@work.com`
Adding this flag means they take a NeoMutt pattern, e.g. `~f @work.com`

> Why is there also the condition `folder_or_mbox` included?

Just to limit it to those hooks

> Does the `-noregex` only have to be in effect for folder and mbox hooks?

Yes.  It only makes sense for those hooks, because the patterns tend to refer to folders and emails.

> BTW: looking for 'noregex' here: [guide]

Yes, I haven't updated the web guide in a while (it's a painful process)

**Comment by klartext** [comment]
charset and iconv hooks are handled by ```mutt_parse_charset_iconv_hook()``` (commit 8588bef8dfb62efbfe521edc0892fd32c0d9d82d)

**Comment by klartext** [comment]
The part inside the `TAILQ_FOREACH()` branch of `mutt_parse_hook()` currently is not implemented in `mutt_parse_charset_iconv_hook()`.
Maybe should be added there...
... but adding such code seems to be necessary for all the sub-parsers, if that part is not moved somewhere to the outside of these.
